File: andorid_deepseek/data/crypto/src/main/java/com/example/transcripto/data/crypto/impl/XorCipher.kt
package com.example.transcripto.data.crypto.impl

import com.example.transcripto.domain.crypto.CipherMethod
import com.example.transcripto.domain.crypto.CipherMethod.CipherParams
import java.util.Base64
import javax.inject.Inject

class XorCipher @Inject constructor() : CipherMethod.Xor() {
    override fun encrypt(input: String, params: CipherParams): String {
        require(params.key.isNotEmpty()) { "Key is required for XOR cipher" }
        val effectiveKey = (params.key + params.salt).toByteArray()
        val inputBytes = input.toByteArray()
        
        val result = inputBytes.mapIndexed { i, byte -> 
            (byte xor effectiveKey[i % effectiveKey.size]).toByte()
        }.toByteArray()
        
        return Base64.getEncoder().encodeToString(result)
    }

    override fun decrypt(input: String, params: CipherParams): String {
        require(params.key.isNotEmpty()) { "Key is required for XOR cipher" }
        val effectiveKey = (params.key + params.salt).toByteArray()
        val inputBytes = Base64.getDecoder().decode(input)
        
        val result = inputBytes.mapIndexed { i, byte -> 
            (byte xor effectiveKey[i % effectiveKey.size]).toByte()
        }.toByteArray()
        
        return String(result)
    }
}

File: android_flutter_claude/transcripto/packages/data_crypto/lib/src/ciphers/xor_cipher.dart
import 'dart:convert';
import 'package:core_common/core_common.dart';
import 'package:domain/domain.dart';

/// XOR cipher implementation
class XorCipher {
  const XorCipher();
  
  /// Encrypts text using XOR cipher (returns Base64 encoded result)
  String encrypt(String input, CipherParams params) {
    final keyBytes = _getEffectiveKey(params);
    final inputBytes = utf8.encode(input);
    final encryptedBytes = _xorBytes(inputBytes, keyBytes);
    return base64Encode(encryptedBytes);
  }
  
  /// Decrypts text using XOR cipher (expects Base64 encoded input)
  String decrypt(String input, CipherParams params) {
    try {
      final keyBytes = _getEffectiveKey(params);
      final encryptedBytes = base64Decode(input);
      final decryptedBytes = _xorBytes(encryptedBytes, keyBytes);
      return utf8.decode(decryptedBytes);
    } catch (e) {
      throw CipherException('Invalid XOR input format: $e');
    }
  }
  
  List<int> _getEffectiveKey(CipherParams params) {
    if (params.key == null || params.key!.trim().isEmpty) {
      throw InvalidKeyException('XOR cipher requires a key');
    }
    
    String effectiveKey = params.key!.trim();
    
    // Add salt to key if enabled
    if (params.useSalt && params.salt != null) {
      final saltString = utf8.decode(params.salt!);
      effectiveKey = effectiveKey + saltString;
    }
    
    return utf8.encode(effectiveKey);
  }
  
  List<int> _xorBytes(List<int> data, List<int> key) {
    if (key.isEmpty) {
      throw InvalidKeyException('XOR key cannot be empty');
    }
    
    final result = <int>[];
    for (int i = 0; i < data.length; i++) {
      final keyByte = key[i % key.length];
      result.add(data[i] ^ keyByte);
    }
    return result;
  }
}
